internal/mongodb/mongodb.go

```diff
@@ -0,0 +1,334 @@
+	// This mechanism is available only in MongoDB Enterprise.
+	//
+	// [https://www.mongodb.com/docs/manual/core/authentication/#std-label-security-auth-kerberos]
+	MechanismGSSAPI Mechanism = "GSSAPI"
+)
+
+type User struct {
+	ID         string            `bson:"_id"`
+	UserID     primitive.Binary  `bson:"userId"`
+	User       string            `bson:"user"`
+	DB         string            `bson:"db"`
+	CustomData map[string]string `bson:"customData"`
+	Roles      []RoleDBRef       `bson:"roles"`
+	Mechanisms []Mechanism       `bson:"mechanisms"`
+}
+
+func (c *Client) ListDBUsers(ctx context.Context, dbName string, filter any) ([]User, error) {
+	query := bson.D{
+		{Key: "usersInfo", Value: 1}, // 1 = all in current db
+	}
+	if filter != nil {
+		query = append(query, bson.E{Key: "filter", Value: filter})
+	}
+	return c.listUsers(ctx, dbName, query)
+}
+
+func (c *Client) ListAllUsers(ctx context.Context, filter any) ([]User, error) {
+	query := bson.D{
+		{Key: "usersInfo", Value: bson.D{
+			{Key: "forAllDBs", Value: true},
+		}},
+	}
+	if filter != nil {
+		query = append(query, bson.E{Key: "filter", Value: filter})
+	}
+	return c.listUsers(ctx, "admin", query)
+}
+
+func (c *Client) GetDBUser(ctx context.Context, dbName, userName string) (User, error) {
+	query := bson.D{
+		{Key: "usersInfo", Value: userName},
+	}
+	users, err := c.listUsers(ctx, dbName, query)
+	if err != nil {
+		return User{}, err
+	}
+	if len(users) == 0 {
+		return User{}, ErrNotFound
+	}
+	return users[0], nil
+}
+
+func (c *Client) listUsers(ctx context.Context, dbName string, query bson.D) ([]User, error) {
+	client, err := c.connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+	db := client.Database(dbName)
+
+	result := db.RunCommand(ctx, query)
+	if err := result.Err(); err != nil {
+		return nil, err
+	}
+	var response struct {
+		CommandResponse `bson:",inline"`
+		Users           []User `bson:"users"`
+	}
+	if err := result.Decode(&response); err != nil {
+		return nil, err
+	}
+	if err := validateResponse(response.CommandResponse); err != nil {
+		return nil, err
+	}
+	return response.Users, nil
+}
+
+type NewUser struct {
+	User       string            `bson:"createUser"`
+	Password   string            `bson:"pwd"`
+	CustomData map[string]string `bson:"customData,omitempty"`
+	Roles      []RoleRef         `bson:"roles"`
+	Mechanisms []Mechanism       `bson:"mechanisms,omitempty"`
+}
+
+func (c *Client) CreateDBUser(ctx context.Context, dbName string, newUser NewUser) (User, error) {
+	if err := c.createDBUser(ctx, dbName, newUser); err != nil {
+		return User{}, err
+	}
+	user, err := c.GetDBUser(ctx, dbName, newUser.User)
+	if err != nil {
+		return User{}, fmt.Errorf("get created user: %w", err)
+	}
+	return user, nil
+}
+
+func (c *Client) createDBUser(ctx context.Context, dbName string, newUser NewUser) error {
```

Comment:
This pattern is confusing to me. Internally, I wouldn't know at first which function to use. What's the purpose of separation here? Do they have the same responsibility?

Reply:
The unpublished/lowercase one is only creating the user without returning it, as the `createUser` MongoDB command doesn't return the created user. I wanted to have "just call the `createUser` command" as a separate function, to encapsulate the context timeout and other local variables.

I don't know how to name it to signify this.

Any suggestions? I can only come up with weird names like:

- `createButDontGetDBUser`
- `onlyCreateDBUser`
- `runCreateUserCommand`

Maybe we could rename the published method? Like:

- `CreateAndGetDBUser`
- Not having both in same method, force caller to use both `CreateDBUser` and then `GetDBUser` explicitly

Reply:
Any suggestions on the name? Personally not super fond of `CreateAndGetDBUser`

Reply:
Updated in bc3e325, wdyt?

Reply:
Personally, I liked more the CreateAndGetDBUser option since it describes exactly what the function does.

Reply:
I don't like having the term "and" inside a method or variable name. Maybe something like "NewDBUser" would be more appropriate? As "New" does suggest something is created, and it returns the created object.

Though the term "New" is conventionally just a constructor in Go, so you'd expect a "New-" function to not have side effects... Hmm! :thinking:

Considering you said the naming was not a blocker, then I'm resolving